src/executor/mod.rs

```diff
@@ -86,115 +76,73 @@ impl<D: Deliverable> ExecutorHandle<D> {
+            client,
+            transaction_timeout,
+        };
+
+        let join_handle = tokio::spawn(executor.run());
+
+        Ok(ExecutorHandle {
+            transaction_counter: weak_counter,
+            worker_counter: Counter::new(),
+            max_transactions: config.max_transactions_per_worker,
+            sender: tx,
+            join_handle,
+        })
     }
-}
 
-impl<D: Deliverable, C: 'static + Connect> Future for Executor<D, C> {
-    type Item = ();
-    type Error = ();
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        loop {
-            // If self.state is not set, then it will be finished
-            // so should only be not set if Finished
-            let state = mem::replace(&mut self.state, ExecutorState::Finished);
-            let mut state_changed = false;
-
-            self.state = match state {
-                ExecutorState::Running(mut receiver) => {
-                    loop {
-                        match receiver.poll() {
-                            Ok(Async::Ready(Some((transaction, counter)))) => {
-                                trace!("Executor: spawning transaction.");
-
-                                transaction.spawn_request(
-                                    &self.client,
-                                    &self.handle,
-                                    self.transaction_timeout.clone(),
-                                    counter,
-                                );
-                            }
-                            // No messages
-                            Ok(Async::NotReady) => break ExecutorState::Running(receiver),
-                            // All senders dropped or errored
-                            // (shouldn't be possible with () error type), shutdown
-                            Ok(Async::Ready(None)) | Err(()) => {
-                                state_changed = true;
-                                break ExecutorState::Draining;
-                            }
-                        }
-                    }
-                }
-                ExecutorState::Draining => {
-                    if self.transaction_counter.count() > 0 {
-                        ExecutorState::Draining
-                    } else {
-                        return Ok(Async::Ready(()));
-                    }
-                }
-                ExecutorState::Finished => panic!("Should not poll() after Executor is finished!"),
-            };
-
-            if !state_changed {
-                break;
-            }
+    async fn run(mut self) {
+        while let Some((transaction, counter)) = self.receiver.next().await {
+            trace!("Executor: spawning transaction.");
+
+            transaction.spawn_request(
+                Arc::clone(&self.client),
+                self.transaction_timeout.clone(),
+                counter,
+            );
+        }
+
+        while self.transaction_counter.count() > 0 {
```

Comment:
Add comment - we only reach this point of the receiver is closed which indicates shutdown or error, we attempt shutdown in either case